File: qtile/.config/qtile/core/rules.py
from libqtile import hook, qtile
from libqtile.config import Group, Key, Match
from libqtile.lazy import lazy

from .keymaps import keys, mod
from .scratchpads import scratchpad

groups = [
    Group(
        "1",
        label="",
        # label="",
        matches=[
            Match(wm_class="Alacritty"),
            Match(wm_class="kitty"),
            Match(wm_class="konsole"),
            Match(wm_class="st-256color"),
            Match(wm_class="org.wezfurlong.wezterm"),
        ],
    ),
    Group(
        "2",
        label="",
        # label="",
        matches=[
            Match(wm_class="Geany"),
            Match(wm_class="editor_cli"),
            Match(wm_class="code-oss"),
            Match(wm_class="jetbrains-idea-ce"),
            Match(wm_class="jetbrains-pycharm"),
            Match(wm_class="jetbrains-dataspell"),
            Match(wm_class="Emacs"),
            Match(wm_class="jupyter-qtconsole"),
            Match(wm_class="PacketTracer"),
        ],
    ),
    Group(
        "3",
        label="󰉋",
        # label="",
        matches=[
            Match(wm_class="felix"),
            Match(wm_class="Thunar"),
            Match(wm_class="Pcmanfm"),
            Match(wm_class="qbittorrent"),
        ],
    ),
    Group(
        "4",
        label="󰊯",
        # label="",
        matches=[
            Match(wm_class="firefox"),
            Match(wm_class="chromium"),
            Match(wm_class="nyxt"),
            Match(wm_class="vieb"),
        ],
    ),
    Group(
        "5",
        label="",
        # label="",
        matches=[
            Match(title="GNU Image Manipulation Program"),  # gimp
            Match(wm_class="vlc"),
            Match(wm_class="obs"),
            Match(wm_class="mpv"),
        ],
    ),
    Group(
        "6",
        label="󰗚",
        # label="",
        matches=[
            Match(wm_class="calibre-gui"),
            Match(wm_class="DesktopEditors"),
            Match(wm_class="org.pwmt.zathura"),
            Match(wm_class="sioyek"),
            Match(wm_class="com.github.johnfactotum.Foliate"),
        ],
    ),
    Group(
        "7",
        label="󰚢",
        # label="",
        matches=[
            Match(wm_class="telegram-desktop"),
        ],
    ),
    Group(
        "8",
        label="󰊖",
        # label="",
        matches=[
            Match(wm_class="Ryujinx"),
            Match(wm_class="yuzu"),
            Match(wm_class="retroarch"),
        ],
    ),
    Group(
        "9",
        label="󰩮",
        # label="",
        matches=[
            Match(wm_class="Gparted"),
            Match(wm_class="xfce4-power-manager-settings"),
            Match(wm_class="Lxappearance"),
            Match(wm_class="Virt-manager"),
        ],
    ),
]

for i in groups:
    keys.extend(
        [
            # mod1 + letter of group = switch to group
            Key(
                [mod],
                i.name,
                lazy.group[i.name].toscreen(),
                desc="Switch to group {}".format(i.name),
            ),
            Key(
                [mod, "shift"],
                i.name,
                lazy.window.togroup(i.name, switch_group=True),
                desc="Switch to and move focused window to group {}".format(i.name),
            ),
        ]
    )

# add scratchpads in the group
# can't add it directly because of the upper `for` statement
groups.append(scratchpad)


# switch to a window's assigned group when it is opened
@hook.subscribe.client_managed
def auto_switch(window):
    if window.group.name != qtile.current_group.name:
        window.group.cmd_toscreen()

File: awesome/.config/awesome/core/rules.lua
-- Standard awesome library
local awful = require("awful")
local gears = require("gears")
require("awful.autofocus")
-- Declarative object management
local ruled = require("ruled")

local popup_tasklist = require("widgets.popup_tasklist")
local c_count_table = require("widgets.client_count")
local systray = require("widgets.tray")


-- {{{ Rules
-- Rules to apply to new clients.
ruled.client.connect_signal(
  "request::rules",
  function()
    -- All clients will match this rule.
    ruled.client.append_rule {
      id = "global",
      rule = {},
      properties = {
        focus = awful.client.focus.filter,
        raise = true,
        screen = awful.screen.preferred,
        placement = awful.placement.no_overlap + awful.placement.no_offscreen
      },
      callback = awful.client.setslave -- Start windows as slave
    }

    -- Floating clients.
    ruled.client.append_rule {
      id = "floating",
      rule_any = {
        instance = {
          "copyq",
          "pinentry"
        },
        class = {
          "Arandr",
          "Blueman-manager",
          "Kruler",
          "Sxiv",
          "Tor Browser",
          "Wpa_gui",
          "veromix",
          "xtightvncviewer",
          "Gpick",
          "Alacritty",
          --"wezterm",
          --"kitty",
          "Lxappearance",
          "Xfce-polkit",
          "Protonvpn",
        },
        -- Note that the name property shown in xprop might be set slightly after creation of the client
        -- and the name shown there might not match defined rules here.
        name = {
          "Event Tester", -- xev.
        },
        role = {
          "AlarmWindow",   -- Thunderbird's calendar.
          "ConfigManager", -- Thunderbird's about:config.
          "pop-up",        -- e.g. Google Chrome's (detached) Developer Tools.
        }
      },
      properties = {
        floating = true,
        placement = awful.placement.centered
      },
    }

    -- Add titlebars to normal clients and dialogs
    ruled.client.append_rule
    {
      id = "titlebars",
      rule_any = {
        type = {
          "normal",
          "dialog"
        }
      },
      properties = {
        titlebars_enabled = true,
      },
    }

    -- app rules | tag rules
    ruled.client.append_rule
    {
      rule_any = {
        class = {
          --"org.wezfurlong.wezterm",
          --"Alacritty",
          "kitty",
          "konsole",
          "st-256color",
        }
      },
      properties = {
        tag = function(c) return c.screen.tags[1] end,
        switch_to_tags = true,
        maximized = false,
        placement = awful.placement.centered,
      }
    }

    ruled.client.append_rule
    {
      rule_any = {
        class = {
          "code-oss",
          "Emacs",
          "Geany",
          "jetbrains-idea-ce",
          "jetbrains-pycharm",
          "jetbrains-dataspell",
        }
      },
      properties = {
        tag = function(c) return c.screen.tags[2] end,
        switch_to_tags = true,
        maximized = false,
        placement = awful.placement.centered,
      }
    }

    ruled.client.append_rule
    {
      rule_any = {
        class = {
          "Pcmanfm",
          "Thunar",
          "qBittorrent",
        }
      },
      properties = {
        tag = function(c) return c.screen.tags[3] end,
        switch_to_tags = true,
        maximized = false,
        placement = awful.placement.centered,
      }
    }

    ruled.client.append_rule
    {
      rule_any = {
        class = {
          "firefox",
          "Chromium",
          "Vieb",
          "Nyxt",
        }
      },
      properties = {
        tag = function(c) return c.screen.tags[4] end,
        switch_to_tags = true,
        maximized = false,
        placement = awful.placement.centered,
      }
    }

    ruled.client.append_rule
    {
      rule_any = {
        class = {
          "Gimp",
          "obs",
          "vlc",
          "mpv",
        }
      },
      properties = {
        tag = function(c) return c.screen.tags[5] end,
        switch_to_tags = true,
        maximized = false,
        placement = awful.placement.centered,
      }
    }

    ruled.client.append_rule
    {
      rule_any = {
        class = {
          "calibre",
          "Zathura",
          "sioyek",
          "DesktopEditors",
        }
      },
      properties = {
        tag = function(c) return c.screen.tags[6] end,
        switch_to_tags = true,
        maximized = false,
        placement = awful.placement.centered,
      }
    }

    ruled.client.append_rule
    {
      rule_any = {
        class = {
          "KotatogramDesktop",
          "TelegramDesktop",
        }
      },
      properties = {
        tag = function(c) return c.screen.tags[7] end,
        switch_to_tags = true,
        maximized = false,
        placement = awful.placement.centered,
      }
    }

    ruled.client.append_rule
    {
      rule_any = {
        class = {
          "Ryujinx",
          "yuzu",
          "retroarch",
          "steamwebhelper",
          "steam",
        }
      },
      properties = {
        tag = function(c) return c.screen.tags[8] end,
        switch_to_tags = true,
        maximized = false,
        placement = awful.placement.centered,
      }
    }

    ruled.client.append_rule
    {
      rule_any = {
        class = {
          "GParted",
          "Xfce4-power-manager-settings",
          "Lxappearance",
          "Virt-manager",
        }
      },
      properties = {
        tag = function(c) return c.screen.tags[9] end,
        switch_to_tags = true,
        maximized = false,
        placement = awful.placement.centered,
      }
    }
  end
)

-- Enable sloppy focus, so that focus follows mouse.
client.connect_signal(
  "mouse::enter",
  function(c)
    c:activate {
      context = "mouse_enter",
      raise = false,
    }
  end
)

-- enable popup tasklist widget with auto-hide
client.connect_signal(
  "focus",
  function()
    if not popup_tasklist.check_visibility() then
      popup_tasklist.toggle_visibility() -- show popup on focus change
      popup_tasklist.auto_hide(2)        -- hides the popup after 2 seconds
    end
  end
)

client.connect_signal(
  "focus",
  -- "property::selected",
  function()
    local current_tag = awful.screen.focused().selected_tag
    local c_count = #current_tag:clients()

    -- c_count_table.client_count = c_count
    c_count_table.update(c_count)
  end
)


-- gears.timer {
--   timeout   = 1, -- Time between checks
--   call_now  = true,
--   autostart = true,
--   callback  = function()
--     if awesome.systray() == 0 then -- if systray has no elements
--       systray.toggle_visibility()
--     end
--   end
-- }
-- }}}
